### PR TITLE
GDB-6528: Scheduled GraphDB backups

### PR DESCRIPTION
**Pull request by mihailradkov**
Added CronJob for scheduling GraphDB backups. The CronJob supports both local and cloud backups.

- Added new configurations under `backup`: `backup.enabled` for toggling the backup CronJob, `backup.type` for selecting between local and cloud and more.
- Local backups support saving the GraphDB backup archives in volume from an existing persistent volume claim, configured with `backup.local`
- Cloud backups support uploading the GraphDB backup archives in one of the supported cloud object storage services, configured with `backup.cloud`

PRO-444: Bumped the chart version to 11.2.0

### CODE REVIEW

File: examples/backup-local/README.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# Local GraphDB Backups Example
+
+This folder contains a simple example of how to configure the Helm chart with local GraphDB backups backed by an existing Persistence Volume
```
Comment (yaskoo) [inline examples/backup-local/README.md:3]:
```suggestion
This example shows how to configure the Helm chart with local GraphDB backups backed by an existing Persistence Volume
```

File: examples/backup-local/README.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+Claim.
+
+## Configuration
+
+The default [values.yaml](values.yaml) in the example uses a simple PVC that should be sufficient for demonstration purposes,
+However, ideally, you should consider using a PVC with cross-zone or cross-region data replication for better resiliency.
```
Comment (yaskoo) [inline examples/backup-local/README.md:9]: Either "however" or "ideally".

File: examples/backup-local/README.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
+# Local GraphDB Backups Example
+
+This folder contains a simple example of how to configure the Helm chart with local GraphDB backups backed by an existing Persistence Volume
+Claim.
+
+## Configuration
+
+The default [values.yaml](values.yaml) in the example uses a simple PVC that should be sufficient for demonstration purposes,
```
Comment (yaskoo) [inline examples/backup-local/README.md:8]:
```suggestion
The default [values.yaml](values.yaml) in the example uses a simple PVC that should be sufficient for demonstration purposes.
```

File: examples/backup-local/README.md [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+However, ideally, you should consider using a PVC with cross-zone or cross-region data replication for better resiliency.
+
+Have in mind that the chart does not rotate backups, this is up to you and your retention policy.
```
Comment (yaskoo) [inline examples/backup-local/README.md:11]: I always find "have in mind" a bit awkward. I think keep / bear in mind are more common.

File: files/scripts/graphdb.sh [hunk 5]
```diff
@@ -4,6 +4,13 @@ set -o errexit
 set -o nounset
 set -o pipefail
 
+function log {
+    local message="$1"
+    local timestamp
+    timestamp=$(date +"%Y-%m-%d %H:%M:%S")
```
Comment (yaskoo) [inline files/scripts/graphdb.sh:10]: why not in the declaration above?

File: files/scripts/graphdb.sh [hunk 6]
```diff
@@ -98,4 +105,87 @@ function createRepositoryFromFile {
...
+
+  local backup_options=
+  backup_options=$(interpolate < "$1")
+
+  log "Creating cloud backup ${BACKUP_NAME}"
+
+  local response=
+  local response_status
+  response=$(mktemp)
+  response_status=$(curl -X POST \
```
Comment (yaskoo) [inline files/scripts/graphdb.sh:139]: post is implied when you give a request body

File: files/scripts/graphdb.sh [hunk 7]
```diff
@@ -98,4 +105,87 @@ function createRepositoryFromFile {
...
+    -isSL \
+    -o "${response}" \
+    -w "Status=%{response_code}" \
+    --header "Authorization: Basic ${GRAPHDB_AUTH_TOKEN}" \
+    --header 'Content-Type: application/json' \
+    --header 'Accept: application/json' \
+    --data-binary "${backup_options}" \
+    --url "http://${GRAPHDB_SERVICE_NAME}:${GRAPHDB_SERVICE_PORT}/rest/recovery/cloud-backup")
+
+  if echo "${response_status}" | grep -q 'Status=200' ; then
```
Comment (yaskoo) [inline files/scripts/graphdb.sh:149]: I've always hated if/elses when they can be avoided - can we invert the check and `return 1` or `exit 1` instead?

File: files/scripts/graphdb.sh [hunk 8]
```diff
@@ -98,4 +105,87 @@ function createRepositoryFromFile {
...
+      output=${output//\$\{$var\}/$value}
+    fi
+  done
+  echo "${output}"
+}
+
+function cloudBackup {
+  BACKUP_TIMESTAMP="$(date +'%Y-%m-%d_%H-%M-%S')"
+  BACKUP_NAME="graphdb-backup-${BACKUP_TIMESTAMP}.tar"
+  export BACKUP_NAME BACKUP_TIMESTAMP
```
Comment (yaskoo) [inline files/scripts/graphdb.sh:129]: why are these exported?

File: templates/jobs/cronjob-backup.yaml [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+                  mountPath: /tmp/utils
+                - name: backup-options
+                  mountPath: /tmp/backup_options.json
+                  subPath: {{ .Values.backup.optionsSecret.secretKey }}
+                {{- if eq .Values.backup.type "local" }}
+                - name: local-backups
+                  mountPath: {{ .Values.backup.local.mountPath }}
+                {{- end }}
+              workingDir: /tmp
+              command: [ "bash" ]
```
Comment (yaskoo) [inline templates/jobs/cronjob-backup.yaml:118]: We can have a function in the script that will switch on the configured type. This way we'll avoid some complexity in the templates

File: files/scripts/graphdb.sh (hunk 5 above)
Comment (mihailradkov) [reply files/scripts/graphdb.sh:10]: It's a best practice when defining a `local`, according to my intellij plugin which cites https://github.com/koalaman/shellcheck/wiki/SC2155

File: files/scripts/graphdb.sh (hunk 6 above)
Comment (mihailradkov) [reply files/scripts/graphdb.sh:139]: I prefer some things to be explicit, like in this case. It's easier to read and understand that this is a POST before even getting to `--data-binary`.

File: files/scripts/graphdb.sh (hunk 8 above)
Comment (mihailradkov) [reply files/scripts/graphdb.sh:129]: Leftovers from when I tried using `envsubst` (which required vars to be exported)

The current solution should work without exporting them, so I'll remove it.

File: templates/jobs/cronjob-backup.yaml (hunk 9 above)
Comment (mihailradkov) [reply templates/jobs/cronjob-backup.yaml:118]: I thought about it too but this effectively would move the complexity... to a template somewhere that will contain the same thing but will be harder to track and find.

It's not a complex template so it's not worth it.

File: files/scripts/graphdb.sh (hunk 8 above)
Comment (yaskoo) [reply files/scripts/graphdb.sh]: btw, `envsubst` should work like this (no export required):
```shell
FOO="bar" \
BAZ="biz" \
envsubst ...
```

**Review by yaskoo — APPROVED** [review]

**Pull request merged**